# **MÓDULO 24 - Bancos de Dados + SQL**

Para nossa atividade inicial de SQL trabalharemos com uma base de vendas.
Abaixo temos o ambiente de preparo para trabalharmos com o SQL lite.

In [1]:
import sqlite3 #Importando a biblioteca necessária
import pandas as pd

In [2]:
df_vendas = pd.read_csv("/content/TB_VENDAS_TAREFA.csv", delimiter=';')

In [3]:
conn = sqlite3.connect(':memory:')
df_vendas.to_sql('tb_vendas', conn, index=False, if_exists='replace')

24

In [4]:
def run_query(query):
    return pd.read_sql_query(query, conn)

# 1) Faça uma consulta que retorne todos os dados da nossa tabela de vendas, com todas linhas e todas colunas.


In [9]:
query = "SELECT * FROM tb_vendas"
result_df = run_query(query)
print(result_df)

    ID_COMPRA  ID_CLIENTE   PRODUTO VALOR_UNID  UNIDADES
0        1247        9081  CAMISETA       34,5         2
1        1248        9081     CALCA       89,9         1
2        1250        9560  CAMISETA       34,5         1
3        1249        9630   BERMUDA       79,5         2
4        1251        9425  CAMISETA       34,5         3
5        1252        9425   VESTIDO        102         1
6        1253        9481     CALCA       89,9         2
7        1254        9785      MEIA       25,9         1
8        1255        9786     CALCA       89,9         2
9        1256        9787   VESTIDO        102         1
10       1257        9788     CALCA       89,9         2
11       1258        9789  CAMISETA       34,5         3
12       1259        9790  CAMISETA       34,5         1
13       1260        9791   BERMUDA       79,5         2
14       1261        9792  CAMISETA       34,5         2
15       1262        9793     CALCA       89,9         2
16       1263        9794     C

# 2) Faça uma consulta que retorne apenas 10 linhas da coluna produto.


In [10]:
query = "SELECT PRODUTO FROM tb_vendas LIMIT 10"
result_df = run_query(query)
print(result_df)

    PRODUTO
0  CAMISETA
1     CALCA
2  CAMISETA
3   BERMUDA
4  CAMISETA
5   VESTIDO
6     CALCA
7      MEIA
8     CALCA
9   VESTIDO


# 3) Faça uma consulta que retorne a média do valor por unidade e também das unidades vendidas. Não se esqueça de renomear o nome das colunas ao cria-las.

In [12]:
query = "SELECT AVG (VALOR_UNID) FROM tb_vendas"
result_df = run_query(query)
print(result_df)

   AVG (VALOR_UNID)
0         63.416667


In [13]:
query = "SELECT AVG (UNIDADES) FROM tb_vendas"
result_df = run_query(query)
print(result_df)

   AVG (UNIDADES)
0        1.833333


###Média do valor = 63.41
###Média de unid = 1.83

# 4) Note que na nossa tabela não temos o total de valor gasto por compra, temos o valor do produto e a quantidade vendida, para termos o valor total precisamos multiplicar essas duas colunas e para multiplicar colunas usamos o *.
# Faça uma consulta que retorne a coluna ID_COMPRA, ID_CLIENTE e o valor total gasto.

Ex: SELECT TICKET_MEDIO*UNIDADE FROM TB_EXEMPLO

In [14]:
query = "SELECT ID_COMPRA, ID_CLIENTE, VALOR_UNID * UNIDADES AS VALOR_TOTAL_GASTO FROM tb_vendas"
result_df = run_query(query)
print(result_df)

    ID_COMPRA  ID_CLIENTE  VALOR_TOTAL_GASTO
0        1247        9081                 68
1        1248        9081                 89
2        1250        9560                 34
3        1249        9630                158
4        1251        9425                102
5        1252        9425                102
6        1253        9481                178
7        1254        9785                 25
8        1255        9786                178
9        1256        9787                102
10       1257        9788                178
11       1258        9789                102
12       1259        9790                 34
13       1260        9791                158
14       1261        9792                 68
15       1262        9793                178
16       1263        9794                 89
17       1264        9795                158
18       1265        9796                136
19       1266        9797                 34
20       1267        9798                204
21       1

# 5) [DESAFIO] Faça uma consulta que traz a média do valor total gasto que foi calculado no exercício anterior. Note que não queremos a média por pedido e sim a média total, apenas essa informação deve vir na consulta.

In [16]:
query = "SELECT AVG(VALOR_UNID * UNIDADES) AS MEDIA_TOTAL_GASTO FROM tb_vendas"
result_df = run_query(query)
# Exibir o resultado
print(result_df)

   MEDIA_TOTAL_GASTO
0            108.625


# 6) Diferencie com suas palavras, considerando o repertório que tem criado da área de dados, a utilidade do SQL e do Python para ciência de dados.
Minimo 2 linhas.

####O SQL e o Python se mostram extremamente necessários para a ciência de dados principalmente para a transformação e pesquisa dentro dos dados, por meio da busca que realizamos dentro dos dados podemos retirar informações valiosas. No início deste projeto tínhamos apenas as colunas presentes dentro do banco de dados e agora já possuímos a média de valores, a média de unidades e a média do total gasto pelos clientes, o que já nos dá informações importantes para levarmos adiante.

#7) Diferencie com suas palavras banco de dados relacionais de não relacionais.
Minimo 3 linhas.

####Os bancos relacionais são o modelo tradicional, com base em meios tabulares, onde dependem de linhas e colunas como é por exemplo o Excel. A sua estrutura possui um esquema fixo e rígido, ele escala de forma vertical onde aumenta a capacidade de mais recursos.
####Os bancos não relacionais abrangem uma variedade de modelos de dados desenvolvidos para lidar com grandes volumes de dados com estruturas variáveis e a necessidade de escalabilidade horizontal. Sua estrutura é de um esquema dinâmico e flexivel, os dados não precisam encaixar em uma estrutura rígida. Diferentes tipos de dados podem ser armazenados no mesmo lugar

#8) Queremos saber da sua opinião, acredita que o SQL pode ser mais prático que o Python para consultas rápidas? Justifique.
Muitas vezes durante a carreia você terá que escolher qual linguagem utilizar para cada momento afim de entregar o que é solicitado o mais rápido e com qualidade possível, exercite sua argumentação = )

Minimo 3 linhas.

####Acredito que sim, existem casos e casos e o SQL é mais simples para as consultas rápidas, não que o python não seja, porém é mais robusto. Tendo em mente que o SQL se apresenta de uma forma mais focada na consulta para insights e informações valiosas de forma rápida, podemos ver que apesar de serem diferentes não possuem realidades tão adversas uma a outra quanto a Python e SQL.